text/0052-resource-importing-support.md

```diff
@@ -0,0 +1,272 @@
+means, into CDK stacks.
+
+## Working Backwards - README
+
+```
+cdk import [STACK]
+
+Import existing resource(s) into the given STACK
+
+Options:
+[... standard CDK CLI and CDK CLI `deploy` options ...]
+  -o, --create-resource-mapping  If specified, CDK will generate a mapping of
+                                 existing physical resources to CDK constructs to
+                                 be imported as. The mapping will be written in
+                                 the given file path. No actual import operation
+                                 will be performed                      [string]
+  -m, --resource-mapping         If specified, CDK will use the given file to
+                                 map physical resources to CDK resources for
+                                 import, instead of interactively asking the
+                                 user. Can be run from scripts          [string]
+  -f, --force                    Allow importing resources with properties
+                                 different from the CDK construct configuration.
+                                 This will most likely lead to a stack drift.
+      --non-interactive          Suppress all interactive prompts. All
+                                 confirmation prompts assume yes, all resources
+                                 that can't be identified will be skipped from
+                                 import. (default: false if stdin is a TTY, true
+                                 otherwise)
+```
+
+Sometimes, it is useful to be able to import (enroll/adopt/...) AWS resources,
+that were created by different means (manually in AWS console, using different
+orchestrator, etc), into a CDK stack. Some resources can simply be deleted and
+recreated by CDK, but for others, this is not convenient: Typically stateful
+resources like S3 Buckets, DynamoDB tables etc, cannot be easily deleted without
+an impact on the service.
+
+To import an existing resource into a CDK stack:
+
+- Run a `cdk diff` to ensure there are no pending changes to the CDK stack you
+  want to import resources into - if there are, apply/discard them first.
+- Add corresponding constructs for the resources to be added in your stack - for
+  example, for an S3 bucket, add something like
+  `new s3.Bucket(this, 'ImportedS3Bucket', {});` - **no other changes must be
+  done to the stack before the import is completed**.
+- Run `cdk import` command - if there are multiple stacks in the CDK app, pass a
+  specific stack name as an argument.
+- If the resource definition contains all information needed for the import,
+  this happens automatically (e.g. an `s3.Bucket` construct has an explicit
+  `bucketName` set), otherwise, CDK will display an interactive prompt to
+  provide neccessary identification information (e.g. the bucket name).
+- After `cdk import` reports success, the resource is managed by CDK. Any
```

Comment:
What are typical ways in which this operation will fail? How are we displaying this to the user?

Reply:
Since I use the existing code for creating/executing the changeset, it will report the error exactly the same way as `cdk deploy`. Typically, the changeset creation fails, I haven't seen the actual execution to fail. If a changeset creation fails, I don't think any further action is needed as no changes were made to the stack yet. And from what I've seen, the CFN error messages are quite clear.

I can imagine a corner case though when a changeset creation passes, but then the physical resource is deleted, before the change set is executed. I am not sure what happens - whether the operation passes and the stack is drifted or it fails. Can try to simulate. Also, there might be different reasons for the changeset execution to fail that I am not aware. The question is if we need to handle those or simply report back to the user, same as with deploy.

Reply:
Probably fine.